docs/design/generic.md

```diff
@@ -0,0 +1,95 @@
+AI models), it is also necessary precursor for implementing support for fetching maven artifacts, which won't be covered
+in this design, but in a followup document.
+
+## Design
+
+In this section, I will try to cover individual parts of the design.
+
+### Source repository
+
+This section will describe the structure of the source repository, that will serve as an input to cachi2. The idea is to
+define a cachi2 lockfile that will specify individual artifacts to fetch along with necessary metadata - e.g. checksums.
+The format chosen for this lockfile is yaml, and will include [purl](https://github.com/package-url/purl-spec) for each
+of the fetched artifacts. This decision was made mainly because it allows for followup implementation of maven support,
+with accurate SBOM information. Here's an example of such a lockfile.
+
+```yaml
+artifacts:
+  - purl: pkg:generic/granite-model?download_url=https://huggingface.co/instructlab/granite-7b-lab/resolve/main/model-00001-of-00003.safetensors?download=true
+    target: granite-model.safetensors
+    checksums:
+      sha256: 07123e1f482356c415f684407a3b8723e10b2cbbc0b8fcd6282c49d37c9c1abc
+```
+
+#### Lockfile format and validation
+
+##### purl (required)
+
+At this point, the only purl type allowed would be `pkg:generic`. This is because cachi2 has no good way of verifying
+additional properties of the fetched artifact that could be included in the resulting SBOM. This should create a strong
+incentive to use this feature in the only truly necessary cases, because it will generate low-quality SBOM components,
+as compared to using other package managers provided by cachi2. Additionally, the only allowed qualifier should be `download_url`.
+
+#### target (optional)
+
+This is mainly for the users convenience, so the files end up in expected locations. Target here means a specific subdirectory
+inside cachi2's output directory. Special care needs to be taken to ensure there is not a conflict with other downloaded files.
+If not specified, filename of the downloaded file will be used.
+
+##### checksums (optional)
+
+I've chosen tho separate checksums from the purl, mostly for better readability of the lockfile, but this can be up for
+discussion. If no checksum is provided, cachi2 should still download the artifact, but report this fact in the output
+SBOM component.
+
+### SBOM
+
+Letting users specify artifacts as purls begs the question of authenticity of the data provided by the users and how it
+should be handled in the resulting SBOM. As described above, the purl is restricted to its basic components, so there is
+very little space for the user to provide inaccurate information. Cachi2 should verify that the file downloaded matches
+checksums and report the purl as-is, as it contains no extra information. The section below outlines how that information
+will be verified at later time.
```

Comment:
Sorry for coming late to the party...

> @eskultety I remember you had a point against having a Cachi2-defined lockfile format?

Yes I did. This project is about dependency processing and generating a record to serve as a proof of contents (or or dependency footprint if you will) for a given project. At the moment it is not about defining any public, long term supported API (besides the CLI). Especially when it comes to formats, given its current simplicity is almost always a bad idea from future perspective due to maintenance burden when it comes to extending it with further structural primitives like grouping or nesting in the future. This design merely mentioned 2 use cases we know about at the moment, but we don't know what the future will hold and with a plethora package managers out there for which the needs are hard to predict I really don't think we want to go down the rabbit hole of maintaining several versions of a generic input lockfile (the current RPM lockfile should serve as a precedent where, luckily, we didn't have to break the structure, but multiple fields had to be added on demand in a fairly short amount of time).

My main reason for not trying to define a format from scratch, however simple and readable it is, is ^this paragraph in the design which just says cachi2 will take the PURL as a given and paste it to the output SBOM. So, to me that bears the question, what exactly is the value of this particular format over an existing CycloneDX? Is it simplicity and readability? Well, the former IMO won't stay true forever and the latter is questionable because the idea is that most lockfiles/dependency lists are generated in an automation manner (i.e. a tool generates them), so it MUST be machine readable, but not necessary human readable, because it's only ever supposed to be an intermediary format for some further processing.
If you look at CycloneDX, while it sure has its flaws, it:
- is a format which is supposed to track artifacts (dependencies in our case)
- has matured over time
- has all the necessary fields we need today and many more we may need in the future
- can be already parsed by cachi2 (in a way) because we can merge our own SBOMs already

Accepting SBOMs on the input for generic artifacts with the same limiting conditions as you outline in this design can be applied to the input SBOM as well. Since we can already do merging, accepting SBOMs feels like the next natural step forward for this project. That said, we could lift the rules and limiting conditions on what we're accepting in the input SBOM gradually, without the need to change the underlying format (CycloneDX) at all.

Reply:
> My main reason for not trying to define a format from scratch, however simple and readable it is, is ^this paragraph in the design which just says cachi2 will take the PURL as a given and paste it to the output SBOM. 

Yes, but the design also mentions that the purl will be very restricted, only to `pkg:generic`, and to `download_url` and `checksums`. Therefore it is taking it as is, but under very restricted conditions.  

> So, to me that bears the question, what exactly is the value of this particular format over an existing CycloneDX? Is it simplicity and readability? Well, the former IMO won't stay true forever and the latter is questionable because the idea is that most lockfiles/dependency lists are generated in an automation manner (i.e. a tool generates them), so it MUST be machine readable, but not necessary human readable, because it's only ever supposed to be an intermediary format for some further processing.

Yes, in traditional package managers, these lockfiles are machine-generated, so they don't have to be user friendly. However, that is because usually there's tooling built around them, so when user needs to update a dependency they just do e.g `poetry add "package==1.1.0"` and don't have to deal with all the specifics of the format. This is **not** the case here. While the format is still machine-readable, there is no tooling to edit it for you. So I would argue that human readability matters here. 

> Accepting SBOMs on the input for generic artifacts with the same limiting conditions as you outline in this design can be applied to the input SBOM as well. Since we can already do merging, accepting SBOMs feels like the next natural step forward for this project. 

My main reason for not wanting to go with SBOM is that it feels like an overkill for this case. And I feel like it isn't a user-friendly format to specify dependencies. 

> That said, we could lift the rules and limiting conditions on what we're accepting in the input SBOM gradually, without the need to change the underlying format (CycloneDX) at all.

While CycloneDX SBOM is flexible and mature, it servers a slightly different purpose in my opinion. We'd still be bending an existing format to our needs. We only need the purl (or its broken-down components), and we'd have to restrict everything else, which in my opinion makes little sense.

Additionally, it poses the same risk as you mentioned, because it might occur that an currently unknown use case appears that CycloneDX won't be able to accommodate, and then you're locked in a specification you _cannot_ change. Custom lockfile format is better for the use case you mentioned.

Reply:
I think I'd still rather keep the simple yaml format for now. If we don't have any expectations of needing to have a full SBOM as an input format, I don't see how the added complexity will help us right now. I also believe we want to keep this feature simple: it should download a file, verify its checksum, and make no other assumptions besides that (that's why it's reported as `pkg:generic`). Allowing this feature to grow beyond this scope might undermine Cachi2's fundamental purpose: accurate SBOMs.

Note that the purl copying from input file to output SBOM is something we're strongly against, and it is not what we're proposing here. If we ever decide to go for a route where, instead of the full source code, we'd receive only an input SBOM to prefetch, there'd be a lot of implications to think of (maybe Cachi2 will never be the project for it).

In any case, this feature will be rolled out initially as experimental, so we have room to change our decision without needing to honor backwards compatibility.

Reply:
> In any case, this feature will be rolled out initially as experimental, so we have room to change our decision without needing to honor backwards compatibility.

How long do you expect to have this as experimental if all initial requirements have been met? That's the thing, you most likely won't have this as experimental the moment you'll need to make a breaking change. The current format doesn't even count with the possibility of being versioned, that is going to be painful. Also, the  moment we start to vendor a new format, I don't think it's appropriate for the format to be version-locked with the project itself, it's possible for sure, but probably should be a different project we maintain. You also need to ensure some backwards compatibility, not expect to break it completely, IOW a particular version of cachi2 must always be able to parse older versions of a lockfile than the version that particular release of cachi2 can handle. These are also consideration I'm not seeing anywhere to be discussed in this thread, but fair enough, it seems folks are in favour of this regardless of any considerations, so I will no longer stand in the way.

Reply:
It most certainly would.

Reply:
I think Erik has a good point about the versioning of the lockfile, as much as I'd like to keep this feature simple and generic, we can't be sure that no changes will be required in the future. Should we add a version field to it then?

Reply:
Added versioning to the lockfile.